transports/sftp/src/test/java/org/mule/transport/sftp/SftpConnectionReturnedToPoolOnErrorTestCase.java

```diff
@@ -0,0 +1,177 @@
+import org.mule.api.endpoint.ImmutableEndpoint;
+import org.mule.api.endpoint.InboundEndpoint;
+import org.mule.api.lifecycle.CreateException;
+import org.mule.tck.probe.PollingProber;
+import org.mule.tck.probe.Probe;
+import org.mule.tck.probe.Prober;
+import org.mule.util.concurrent.Latch;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.junit.Test;
+
+public class SftpConnectionReturnedToPoolOnErrorTestCase extends AbstractSftpFunctionalTestCase
+{
+
+    public static final String FILE1_NAME = "file1.txt";
+
+    private static AtomicInteger pollingInvokedTimes;
+    private static String[] availableFiles;
+
+    private static Latch sftpServerStopReq = new Latch();
+    private static Latch sftpServerStopped = new Latch();
+
+    @Override
+    protected String getConfigFile()
+    {
+        return "mule-sftp-connection-returned-to-pool-on-error-config.xml";
+    }
+
+    @Override
+    protected void setUpTestData() throws IOException
+    {
+        sftpClient.storeFile(FILE1_NAME, new ByteArrayInputStream(TEST_MESSAGE.getBytes()));
+    }
+
+    @Override
+    protected void doSetUpBeforeMuleContextCreation() throws Exception
+    {
+        pollingInvokedTimes = new AtomicInteger(0);
+        super.doSetUpBeforeMuleContextCreation();
+
+        Executors.newSingleThreadExecutor().execute(new Runnable()
+        {
+            @Override
+            public void run()
+            {
+                try
+                {
+                    sftpServerStopReq.await();
+                    sftpServer.stop();
+                    setUpServer();
+                    sftpServerStopped.countDown();
+                }
+                catch (InterruptedException e)
+                {
+                    // Nothing to do
+                }
+            }
+        });
+    }
+
+    @Override
+    protected void doTearDownAfterMuleContextDispose() throws Exception
+    {
+        // In case there was a failure in @Before and the client is not set, avoid throwing an NPE that would hide the
+        // first exception.
+        if (sftpClient != null)
+        {
+            cleanUpTestFolder();
+
+            sftpClient.disconnect();
+        }
+    }
+
+    @Test
+    public void testConnectionReturned() throws Exception
+    {
+        SftpConnector c = (SftpConnector) muleContext.getRegistry().lookupConnector("sftpPooledConnector");
+        final int connectorPoolSize = c.getMaxConnectionPoolSize();
+        assertEquals(true, c.useConnectionPool());
+
+        Prober prober = new PollingProber(RECEIVE_TIMEOUT, 50);
+        prober.check(new Probe()
+        {
+            @Override
+            public boolean isSatisfied()
+            {
+                return isSftpServerStopped() && pollingInvokedTimes.get() > connectorPoolSize;
+            }
+
+            protected boolean isSftpServerStopped()
```

Comment:
This would make more sense as "hasSftpServerStopped" since it might actually be running.